### PR TITLE
[Feature-2574][Server]Specify Network Interface

### LINKED ISSUE #2574: [Feature]Configurable registration  ipaddress
**Is your feature request related to a problem? Please describe.**
on macos,OSUtils.getHost() returns diffrent ip in running time. Also,the method OSUtils.getHost() didn't
perform well in container situation like docker ,some error may happen
**Describe the solution you'd like**
provide a param like register-ip to specific the ip which will be registried on zk
**Describe alternatives you've considered**
provide a param like networkinterface to specific the networkinterface  which will be registried on zk
**Additional context**
8C1E5A5B-CC64-40FE-9049-5FC6154E72FE
25DECE0A-2A92-479F-A76F-7DAEB8F691FF
英文不好,简单的说就是可以指定注册到zk上的ip或可选的网卡配置

### PR DESCRIPTION
**Pull request by CalvinKirs**
## What is the purpose of the pull request

This closes #2574

## Verify this pull request

This pull request is already covered by existing tests, such as *(please describe tests)*.

This change added tests and can be verified as follows:

  - *Manually verified the change by testing locally.*

### CODE REVIEW

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java [hunk 1]
```diff
@@ -206,4 +215,8 @@ public static boolean ignoreNetworkInterface(NetworkInterface networkInterface)
                 || !networkInterface.isUp();
     }
 
+    public static boolean isSpecifyNetworkInterface(NetworkInterface networkInterface) {
+        String preferredNetworkInterface = System.getProperty(DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE);
+        return Objects.equals(networkInterface.getDisplayName(), preferredNetworkInterface);
+    }
```
Comment (yangyichao-mango) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java:221]: Hi,

Good job,
`Utils` should be used as a tool class and it is better not to involve business process.
Is it better to implement this part in the business process class?

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java [hunk 2]
```diff
@@ -51,6 +48,8 @@ private NetUtils() {
 
     private static volatile String HOST_ADDRESS;
 
+    private static  String DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE = "dolphin.scheduler.network.interface.preferred";
+
```
Comment (yangyichao-mango) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java:52]: Hi,

Is it better to move `DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE` to `Constants.java`?

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java (hunk 1 above)
Comment (CalvinKirs) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java]: - Hello, this is exclusive to net, other business is not involved, so I think it is better to achieve here,how do you think？

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java (hunk 2 above)
Comment (CalvinKirs) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java]: Thanks for your suggestions, I have completed the changes.

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java (hunk 1 above)
Comment (yangyichao-mango) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java]: > * Hello, this is exclusive to net, other business is not involved, so I think it is better to achieve here,how do you think？

Hi,

I think it is better not to add the `isSpecifyNetworkInterface` in `findNetworkInterface` of `utils`, but I find there is only one call of `findNetworkInterface`, so I don't think of a better place to add this at present.
I think we can think about the implement scheme again later.

**Comment by yangyichao-mango** [comment]
Hi,

I find you use the function of `private static synchronized InetAddress getLocalAddress0()`, can you help to explain will it produce the performance issues when call the `getLocalAddress0()` function.
Thx a lot.

**Comment by CalvinKirs** [comment]
> Hi,
> 
> I find you use the function of `private static synchronized InetAddress getLocalAddress0()`, can you help to explain will it produce the performance issues when call the `getLocalAddress0()` function.
> Thx a lot.

There will be no impact here

**Comment by CalvinKirs** [comment]
done.

**Comment by qiaozhanwei** [comment]
https://lists.apache.org/thread.html/r3442e2bd785628ad91655f9e5cd79c1b6d33bfa7826e70f6070158d7%40%3Cdev.dolphinscheduler.apache.org%3E

This is mail discuss before，please refer

The conclusion at the time ，InetAddress.getLocalHost().getHostAddress() not support docker host mode
but DS also have a utils at org.apache.dolphinscheduler.remote.utils.IPUtils
But IPUtils also have another problem，for vm machine have many ips，netty is not access

your PR can solve this problem ?

**Comment by szs122** [comment]
I think this need modify like this. 
`private static boolean isSpecifyNetworkInterface(NetworkInterface networkInterface) {
        String preferredNetworkInterface = PropertyUtils.getString(DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE);
        logger.info("config name:{}, net work interface: {}, addrs: {}",preferredNetworkInterface,  networkInterface, networkInterface.getInterfaceAddresses());
        return Objects.equals(networkInterface.getName(), preferredNetworkInterface) || Objects.equals(networkInterface.getDisplayName(), preferredNetworkInterface);
    }`
to fix which interface to use.

common.properties add
# Network Interface name
#dolphin.scheduler.network.interface.preferred=eth0

**Comment by szs122** [comment]
`
# Network Interface name
#dolphin.scheduler.network.interface.preferred=eth0
`

**Comment by szs122** [comment]
I think this modify like this

`
    private static boolean isSpecifyNetworkInterface(NetworkInterface networkInterface) {
        String preferredNetworkInterface = PropertyUtils.getString(DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE);
        logger.info("config name:{}, net work interface: {}, addrs: {}",preferredNetworkInterface,  networkInterface, networkInterface.getInterfaceAddresses());
        return Objects.equals(networkInterface.getName(), preferredNetworkInterface) || Objects.equals(networkInterface.getDisplayName(), preferredNetworkInterface);
    }
`

common.properties add interface config

`
# Network Interface name
#dolphin.scheduler.network.interface.preferred=eth0
`

**Comment by szs122** [comment]
I think this modify like this, the name is more better.

```
 private static boolean isSpecifyNetworkInterface(NetworkInterface networkInterface) {
        String preferredNetworkInterface = PropertyUtils.getString(DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE);
        logger.info("config name:{}, net work interface: {}, addrs: {}",
                preferredNetworkInterface,  networkInterface, networkInterface.getInterfaceAddresses());
        return Objects.equals(networkInterface.getName(), preferredNetworkInterface) || 
                Objects.equals(networkInterface.getDisplayName(), preferredNetworkInterface);
    }
```

common.properties add interface config
```
# Network Interface name
#dolphin.scheduler.network.interface.preferred=eth0
```

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java (hunk 1 above)
Comment (szs122) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java]: > I think this modify like this, the name is more better.
> 
> ```
>  private static boolean isSpecifyNetworkInterface(NetworkInterface networkInterface) {
>         String preferredNetworkInterface = PropertyUtils.getString(DOLPHIN_SCHEDULER_PREFERRED_NETWORK_INTERFACE);
>         logger.info("config name:{}, net work interface: {}, addrs: {}",
>                 preferredNetworkInterface,  networkInterface, networkInterface.getInterfaceAddresses());
>         return Objects.equals(networkInterface.getName(), preferredNetworkInterface) || 
>                 Objects.equals(networkInterface.getDisplayName(), preferredNetworkInterface);
>     }
> ```
> 
> common.properties add interface config
> 
> ```
> # Network Interface name
> #dolphin.scheduler.network.interface.preferred=eth0
> ```

**Pull request merged**